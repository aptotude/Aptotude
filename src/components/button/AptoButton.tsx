import React from 'react';
import classNames from 'classnames';
import './aptoButton.scss';
import { StandardTypes } from '../../utils/standardTypes';

export type ButtonKind = 'button' | 'link';
export type ButtonVariant = 'primary' | 'secondary' | 'secondaryDark' | 'danger';

export interface AptoButtonDisplayProps extends StandardTypes {
  kind?: ButtonKind;
  variant?: ButtonVariant;
  active?: boolean;
  disabled?: boolean;
  onClick?: (event: React.MouseEvent<HTMLButtonElement|HTMLAnchorElement>) => void;
  href?: string;
  role?: string;
  title?: string;
  target?: string;
}

function missingHref(href: undefined | string) {
  return !href || href.trim() === '#';
}

const COMPONENT_PREFIX = 'AptoButton';

export class AptoButton extends React.PureComponent<AptoButtonDisplayProps> {
  public static defaultProps = {
    kind: 'button',
    variant: 'primary',
    active: false,
    disabled: false
  };

  public constructor(props = {}) {
    super(props);

    this.handleClick = this.handleClick.bind(this);
  }

  public handleClick(event: React.MouseEvent<HTMLButtonElement|HTMLAnchorElement>) {
    const {disabled, href, onClick } = this.props;

    if (disabled || missingHref(href)) {
      if (event) {
        event.preventDefault();
      }
    }

    if (disabled) {
      if (event) {
        event.stopPropagation();
      }
      return;
    }

    if (onClick) {
      onClick(event);
    }
  }

  public render() {
    const {
      variant,
      kind,
      forwardRef,
      className,
      children,
      disabled,
      active,
      href,
      title,
      ...rest
    } = this.props;

    if (href && href === '#') {
      rest.role = rest.role || 'button';
    }

    const Component: any = href ? 'a' : 'button';

    const classes = classNames(
      COMPONENT_PREFIX,
      `${COMPONENT_PREFIX}--${kind}`,
      `${COMPONENT_PREFIX}--${variant}`,
      disabled && `${COMPONENT_PREFIX}--disabled`,
      active && `${COMPONENT_PREFIX}--active`,
      className
    );

    return (
      <Component
        {...rest}
        ref={forwardRef}
        aria-label={title}
        href={href}
        disabled={disabled || undefined}
        onClick={this.handleClick}
        className={classes}>
          {children}
      </Component>
    );
  }
}
