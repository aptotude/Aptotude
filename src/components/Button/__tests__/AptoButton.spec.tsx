import React from 'react';
import renderer from 'react-test-renderer';
import AptoButton from '../AptoButton';

test('Button renders', () => {
  const component = renderer.create(
    <AptoButton />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Button children render', () => {
  const component = renderer.create(
    <AptoButton>Foo</AptoButton>
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Primary Button renders', () => {
  const component = renderer.create(
    <AptoButton variant="primary" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Secondary Button renders', () => {
  const component = renderer.create(
    <AptoButton variant="secondary" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('SecondaryDark Button renders', () => {
  const component = renderer.create(
    <AptoButton variant="secondaryDark" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Danger Button renders', () => {
  const component = renderer.create(
    <AptoButton variant="danger" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Link button renders', () => {
  const component = renderer.create(
    <AptoButton kind="link" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Button button renders', () => {
  const component = renderer.create(
    <AptoButton kind="button" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Secondary Link button renders', () => {
  const component = renderer.create(
    <AptoButton kind="link" variant="secondary" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Disabled button renders', () => {
  const component = renderer.create(
    <AptoButton disabled />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Active button renders', () => {
  const component = renderer.create(
    <AptoButton active />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Anchor renders with role attribute', () => {
  const component = renderer.create(
    <AptoButton href="#" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Anchor renders without role', () => {
  const component = renderer.create(
    <AptoButton href="http://www.google.com" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Anchor target renders', () => {
  const component = renderer.create(
    <AptoButton href="#" target="_blank" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Disabled anchor renders', () => {
  const component = renderer.create(
    <AptoButton href="#" disabled />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Active anchor renders', () => {
  const component = renderer.create(
    <AptoButton active href="#" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Aria Label attribute renders', () => {
  const component = renderer.create(
    <AptoButton title="some title" />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Button is clickable', () => {
  const onClick = jest.fn();
  const component = renderer.create(
    <AptoButton onClick={onClick}>Click</AptoButton>
  );
  const tree = component.toJSON();
  if (tree) {
    tree.props.onClick();
  }
  expect(onClick).toBeCalled();
});

test('Button is not clickable when disabled', () => {
  const onClick = jest.fn();
  const component = renderer.create(
    <AptoButton onClick={onClick} disabled>Click</AptoButton>
  );
  const tree = component.toJSON();
  if (tree) {
    tree.props.onClick();
  }
  expect(onClick).not.toBeCalled();
});
