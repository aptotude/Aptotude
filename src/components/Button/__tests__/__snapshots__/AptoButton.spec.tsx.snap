// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Active anchor renders 1`] = `
<a
  className="AptoButton AptoButton--button AptoButton--primary AptoButton--active"
  href="#"
  onClick={[Function]}
  role="button"
/>
`;

exports[`Active button renders 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--primary AptoButton--active"
  onClick={[Function]}
/>
`;

exports[`Anchor renders with role attribute 1`] = `
<a
  className="AptoButton AptoButton--button AptoButton--primary"
  href="#"
  onClick={[Function]}
  role="button"
/>
`;

exports[`Anchor renders without role 1`] = `
<a
  className="AptoButton AptoButton--button AptoButton--primary"
  href="http://www.google.com"
  onClick={[Function]}
/>
`;

exports[`Anchor target renders 1`] = `
<a
  className="AptoButton AptoButton--button AptoButton--primary"
  href="#"
  onClick={[Function]}
  role="button"
  target="_blank"
/>
`;

exports[`Aria Label attribute renders 1`] = `
<button
  aria-label="some title"
  className="AptoButton AptoButton--button AptoButton--primary"
  onClick={[Function]}
/>
`;

exports[`Button button renders 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--primary"
  onClick={[Function]}
/>
`;

exports[`Button children render 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--primary"
  onClick={[Function]}
>
  Foo
</button>
`;

exports[`Button renders 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--primary"
  onClick={[Function]}
/>
`;

exports[`Danger Button renders 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--danger"
  onClick={[Function]}
/>
`;

exports[`Disabled anchor renders 1`] = `
<a
  className="AptoButton AptoButton--button AptoButton--primary AptoButton--disabled"
  disabled={true}
  href="#"
  onClick={[Function]}
  role="button"
/>
`;

exports[`Disabled button renders 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--primary AptoButton--disabled"
  disabled={true}
  onClick={[Function]}
/>
`;

exports[`Link button renders 1`] = `
<button
  className="AptoButton AptoButton--link AptoButton--primary"
  onClick={[Function]}
/>
`;

exports[`Primary Button renders 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--primary"
  onClick={[Function]}
/>
`;

exports[`Secondary Button renders 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--secondary"
  onClick={[Function]}
/>
`;

exports[`Secondary Link button renders 1`] = `
<button
  className="AptoButton AptoButton--link AptoButton--secondary"
  onClick={[Function]}
/>
`;

exports[`SecondaryDark Button renders 1`] = `
<button
  className="AptoButton AptoButton--button AptoButton--secondaryDark"
  onClick={[Function]}
/>
`;
